File: engine/src/com/magma/engine/utils/MagmaMath.java
package com.magma.engine.utils;

import com.badlogic.gdx.math.Circle;
import com.badlogic.gdx.math.Intersector;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Shape2D;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.magma.engine.maps.triggers.CustomTrigger;

public class MagmaMath {
	public static Rectangle scaleRectangle(Rectangle rect, float scale) {
		rect.x *= scale;
		rect.y *= scale;
		rect.width *= scale;
		rect.height *= scale;
		return rect;
	}

	public static enum PadAlign {
		LEFT, CENTER, RIGHT
	}

	public static String padString(String input, char padSymbol, int totalLength, PadAlign align) {
		// check if text to short
		if (input.length() >= totalLength) {
			return input;
		}
		
		String result;
		switch (align) {
		case CENTER:
			throw new IllegalArgumentException("Alignment not implemented");
		case RIGHT:
			result = input;
			for (int i = 0; i < totalLength - input.length(); i++) {
				result += padSymbol;
			}
			break;
		case LEFT:
			result = "";
			for (int i = 0; i < totalLength - input.length(); i++) {
				result += padSymbol;
			}
			result += input;
			break;
		default:
			throw new IllegalArgumentException("Alignment not implemented");
		}
		return result;
	}
	

	public static float getDistanceToCircle(float x, float y, Circle circle) {
		return (circle.x - x) * (circle.x - x) + (circle.y - y) * (circle.y - y);
	}
	
	public static boolean shapesOverlap(Shape2D a, Shape2D b) { // this is probably dumb
		if (a instanceof Circle) {
			Circle circle = (Circle) a;
			if (b instanceof Circle) {
				return circle.overlaps((Circle) b);
			}
			if (b instanceof Rectangle) {
				return Intersector.overlaps(circle,(Rectangle) b);
			}
		}
		else if (a instanceof Rectangle) {
			Rectangle rect = (Rectangle) a;
			if (b instanceof Rectangle) {
				return Intersector.overlaps(rect, (Rectangle) b);
			}
			if (b instanceof Circle) {
				return Intersector.overlaps((Circle) b, rect);
				//return rectOverlapsCircle(rect, (Circle) b);
			}
		}
		throw new IllegalArgumentException("Invalid shapes");
	}

	public static Shape2D extractShape(Actor actor) {
		Shape2D shape;
		if (actor instanceof CustomTrigger) {
			shape = ((CustomTrigger) actor).getShape();
		} else {
			shape = new Rectangle(actor.getX(), actor.getY(), actor.getWidth(), actor.getHeight()); // TODO pool?
		}
		return shape;
	}
}

File: engine/src/com/magma/engine/utils/MagmaMath.kt
package com.magma.engine.utils

import com.badlogic.gdx.math.Circle
import com.badlogic.gdx.math.Intersector
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Shape2D
import com.badlogic.gdx.scenes.scene2d.Actor
import com.magma.engine.maps.triggers.CustomTrigger

object MagmaMath {
    fun scaleRectangle(rect: Rectangle, scale: Float): Rectangle {
        rect.x *= scale
        rect.y *= scale
        rect.width *= scale
        rect.height *= scale
        return rect
    }

	fun padString(input: String, padSymbol: Char, totalLength: Int, align: PadAlign?): String {
        // check if text to short
        if (input.length >= totalLength) {
            return input
        }
        var result: String
        when (align) {
            PadAlign.CENTER -> throw IllegalArgumentException("Alignment not implemented")
            PadAlign.RIGHT -> {
                result = input
                var i = 0
                while (i < totalLength - input.length) {
                    result += padSymbol
                    i++
                }
            }

            PadAlign.LEFT -> {
                result = ""
                var i = 0
                while (i < totalLength - input.length) {
                    result += padSymbol
                    i++
                }
                result += input
            }

            else -> throw IllegalArgumentException("Alignment not implemented")
        }
        return result
    }

    fun getDistanceToCircle(x: Float, y: Float, circle: Circle): Float {
        return (circle.x - x) * (circle.x - x) + (circle.y - y) * (circle.y - y)
    }

    fun shapesOverlap(a: Shape2D?, b: Shape2D?): Boolean { // this is probably dumb
        if (a is Circle) {
            val circle = a
            if (b is Circle) {
                return circle.overlaps(b as Circle?)
            }
            if (b is Rectangle) {
                return Intersector.overlaps(circle, b as Rectangle?)
            }
        } else if (a is Rectangle) {
            val rect = a
            if (b is Rectangle) {
                return Intersector.overlaps(rect, b as Rectangle?)
            }
            if (b is Circle) {
                return Intersector.overlaps(b as Circle?, rect)
                //return rectOverlapsCircle(rect, (Circle) b);
            }
        }
        throw IllegalArgumentException("Invalid shapes")
    }

    fun extractShape(actor: Actor): Shape2D {
        val shape: Shape2D = if (actor is CustomTrigger) {
            (actor as CustomTrigger).shape
        } else {
            Rectangle(actor.x, actor.y, actor.width, actor.height) // TODO pool?
        }
        return shape
    }

    enum class PadAlign {
        LEFT, CENTER, RIGHT
    }
}